diff --git a/Syx/Syx.h b/Syx/Syx.h
index 4a30a96..7bee3f3 100644
--- a/Syx/Syx.h
+++ b/Syx/Syx.h
@@ -1,10 +1,15 @@
+#pragma once
 #include <stdexcept>
 #include <vector>
 #include <string>
-#include <sstream>
-#include <Psapi.h>
+#include <Windows/MinWindows.hpp>
 #include <Windows.h>
-
+#include <Psapi.h>
+#ifdef UNICODE
+#define wcharptr_t LPCWSTR
+#else
+#define wcharptr_t LPCSTR
+#endif
 /**
  * SyxLib is a C++ class that provides functions for memory operations and pattern searching in a Windows environment.
  * It simplifies common memory-related tasks such as reading and writing memory, searching for patterns in memory,
@@ -43,11 +48,14 @@ public:
      * @param szModule The name of the module.
      * @return MODULEINFO structure containing information about the module.
      */
-    static MODULEINFO GetModuleInfo(const wchar_t* szModule)
+    static MODULEINFO GetModuleInfo(wcharptr_t szModule)
+    {
+        return GetModuleInfoA(GetModuleHandle(szModule));
+    }
+    static MODULEINFO GetModuleInfoA(HMODULE hModule)
     {
-        MODULEINFO modinfo{ 0 };
-        HMODULE hModule = GetModuleHandle(szModule);
-        if (hModule == 0)
+        MODULEINFO modinfo{ nullptr };
+        if (hModule == nullptr)
             return modinfo;
         GetModuleInformation(GetCurrentProcess(), hModule, &modinfo, sizeof(MODULEINFO));
         return modinfo;
@@ -74,6 +82,8 @@ public:
         return 0;
     }
 
+    // static uintptr_t FindPatternAB(HMODULE module, char* pattern, char* mask);
+    // static uintptr_t FindPatternAC(MODULEINFO, char*, char*);
     /**
      * Searches for a pattern in an auto memory region.
      *
@@ -82,15 +92,23 @@ public:
      * @param mask The mask specifying which bytes in the pattern to consider.
      * @return The address of the found pattern, or NULL if not found.
      */
-    static uintptr_t FindPatternA(const wchar_t* module, char* pattern, char* mask)
+    static uintptr_t FindPatternA(wcharptr_t module, char* pattern, char* mask)
+    {
+        return FindPatternAC(GetModuleInfo(module), pattern, mask);
+    }
+    static uintptr_t FindPatternAB(const HMODULE module, const char* pattern, const char* mask)
+    {
+        return FindPatternAC(GetModuleInfoA(module), pattern, mask);
+    }
+    static uintptr_t FindPatternAC(MODULEINFO mInfo, const char* pattern, const char* mask)
     {
-        MODULEINFO mInfo = GetModuleInfo(module);
         uintptr_t base = (uintptr_t)mInfo.lpBaseOfDll;//10000000
-        uintptr_t size = (uintptr_t)mInfo.SizeOfImage;//6000000
+        uintptr_t size = mInfo.SizeOfImage;//6000000
 
-        uintptr_t patternLength = (uintptr_t)strlen(mask);
+        uintptr_t patternLength = strlen(mask);
+        size_t patternOffset = size > patternLength ? size - patternLength : 0;
 
-        for (uintptr_t i = 0; i < size - patternLength; i++)
+        for (uintptr_t i = 0; i < patternOffset; i++)
         {
             bool found = true;
             for (uintptr_t j = 0; j < patternLength; j++)
@@ -162,7 +180,7 @@ public:
     {
         // Allocate a memory page that is going to contain executable code.
         MEMORY_BASIC_INFORMATION mbi;
-        const SIZE_T pageSize = 0x1000;
+        constexpr SIZE_T pageSize = 0x1000;
         for (SIZE_T addr = reinterpret_cast<SIZE_T>(src); addr > reinterpret_cast<SIZE_T>(src) - 0x80000000; addr = reinterpret_cast<SIZE_T>(mbi.BaseAddress) - 1)
         {
             if (!VirtualQuery(reinterpret_cast<LPCVOID>(addr), &mbi, sizeof(mbi)))
@@ -172,7 +190,7 @@ public:
 
             if (mbi.State == MEM_FREE)
             {
-                if (presenthook64 = reinterpret_cast<HookContext*>(VirtualAlloc(mbi.BaseAddress, pageSize, MEM_RESERVE | MEM_COMMIT, PAGE_EXECUTE_READWRITE)))
+                if ((presenthook64 = reinterpret_cast<HookContext*>(VirtualAlloc(mbi.BaseAddress, pageSize, MEM_RESERVE | MEM_COMMIT, PAGE_EXECUTE_READWRITE))))
                 {
                     break;
                 }
